_source/_posts/2021-11-08-k8s-api-server-oidc.md

```diff
@@ -0,0 +1,565 @@
+
+```hcl
+# Set up OKTA groups
+resource "okta_group" "k8s_admin" {
+  name        = "k8s-admins"
+  description = "Users who can access k8s cluster as admins"
+}
+
+resource "okta_group" "k8s_restricted_users" {
+  name        = "k8s-restricted-users"
+  description = "Users who can only view pods and services in default namespace"
+}
+```
+
+### Assign users to the groups
+
+The below snippet looks up existing users and adds them to the groups. You can add any number of users at this stage, or you can skip adding users and do it via the Okta Admin Console later. For this exercise, I'm fetching existing users. You could also create new users using the `okta_user` resource.
+
+```hcl
+# Assign users to the groups
+data "okta_user" "admin" {
+  search {
+    name  = "profile.email"
+    value = "<your_admin_user_email>"
+  }
+}
+
+resource "okta_group_memberships" "admin_user" {
+  group_id = okta_group.k8s_admin.id
+  users = [
+    data.okta_user.admin.id
+  ]
+}
+
+data "okta_user" "restricted_user" {
+  search {
+    name  = "profile.email"
+    value = "<another_user_email>"
+  }
+}
+
+resource "okta_group_memberships" "restricted_user" {
+  group_id = okta_group.k8s_restricted_users.id
+  users = [
+    data.okta_user.restricted_user.id
+  ]
+}
+```
+
+### Create an OIDC application
+
+Now that our groups are in place let's create an OIDC application. We will set the application type to `native` and use PKCE as client authentication, which is much more secure than using a client secret. We will also set the redirect URIs to `localhost:8000` so that we can work with kubectl locally. We should also assign the groups we created earlier to this application here. Finally, we can capture the client ID of the created app using an output variable.
+
+```hcl
+# Create an OIDC application
+
+resource "okta_app_oauth" "k8s_oidc" {
+  label                      = "k8s OIDC"
+  type                       = "native" # this is important
+  token_endpoint_auth_method = "none"   # this sets the client authentication to PKCE
+  grant_types = [
+    "authorization_code"
+  ]
+  response_types = ["code"]
+  redirect_uris = [
+    "http://localhost:8000",
+  ]
+  post_logout_redirect_uris = [
+    "http://localhost:8000",
+  ]
+  lifecycle {
+    ignore_changes = [groups]
+  }
+}
+
+# Assign groups to the OIDC application
+resource "okta_app_group_assignments" "k8s_oidc_group" {
+  app_id = okta_app_oauth.k8s_oidc.id
+  group {
+    id = okta_group.k8s_admin.id
+  }
+  group {
+    id = okta_group.k8s_restricted_users.id
+  }
+}
+
+output "k8s_oidc_client_id" {
+  value = okta_app_oauth.k8s_oidc.client_id
+}
+```
+
+### Create an authorization server
+
+Next, we would need an authorization server so that we can define policies and claims. We can also capture the issuer URL in an output variable.
```

Comment:
```suggestion
Next, we need an authorization server so that we can define policies and claims. We will also capture the issuer URL in an output variable.
```